batch/src/main/scala/uk/gov/ons/addressindex/utils/SparkProvider.scala

```diff
@@ -37,11 +37,53 @@ object SparkProvider {
   private lazy val sparkContext = SparkContext.getOrCreate(conf)
   lazy val sqlContext = SQLContext.getOrCreate(sparkContext)
 
+  sqlContext.udf.register("concatNag", concatNag(_: String, _: String, _: String, _: String, _: String, _: String, _: String,
+    _: String, _: String, _: String, _: String, _: String, _: String, _: String, _: String))
+
   val incrementalId = new AtomicInteger()
 
   def registerTempTable(dataFrame: DataFrame, name: String): String = {
     val generatedName = name + incrementalId.getAndIncrement()
     dataFrame.registerTempTable(generatedName)
     generatedName
   }
+
+
+  def concatNag(saoStartNumber: String, saoEndNumber: String, saoEndSuffix: String, saoStartSuffix: String,
+                saoText: String, organisation: String, paoStartNumber: String, paoStartSuffix: String,
+                paoEndNumber: String, paoEndSuffix: String, paoText: String, streetDescriptor: String,
+                townName: String, locality: String, postcodeLocator: String): String = {
+
+    val saoLeftRangeExists = saoStartNumber.nonEmpty || saoStartSuffix.nonEmpty
+    val saoRightRangeExists = saoEndNumber.nonEmpty || saoEndSuffix.nonEmpty
+    val saoHyphen = if (saoLeftRangeExists && saoRightRangeExists) "-" else ""
+
+    val saoNumbers = Seq(saoStartNumber, saoStartSuffix, saoHyphen, saoEndNumber, saoEndSuffix)
+      .map(_.trim).mkString
+    val sao =
+      if (saoText == organisation || saoText.isEmpty) saoNumbers
+      else if (saoNumbers.isEmpty) s"$saoText"
+      else s"$saoNumbers $saoText"
+
+    val paoLeftRangeExists = paoStartNumber.nonEmpty || paoStartSuffix.nonEmpty
+    val paoRightRangeExists = paoEndNumber.nonEmpty || paoEndSuffix.nonEmpty
+    val paoHyphen = if (paoLeftRangeExists && paoRightRangeExists) "-" else ""
+
+    val paoNumbers = Seq(paoStartNumber, paoStartSuffix, paoHyphen, paoEndNumber, paoEndSuffix)
+      .map(_.trim).mkString
+    val pao =
+      if (paoText == organisation || paoText.isEmpty) paoNumbers
+      else if (paoNumbers.isEmpty) s"$paoText"
+      else s"$paoText $paoNumbers"
```

Comment:
Is is correct that paoText comes before paoNumbers? I would have assumed that they were the other way around e.g. "65B Victorian House".